pkg/services/object/put/ec.go

```diff
@@ -0,0 +1,135 @@
+			return fmt.Errorf("apply EC rule #%d (%s): %w", i, ecRules[i], err)
+		}
+
+		for j := i + 1; j < len(ecRules); j++ {
+			if ecRules[i] != ecRules[j] {
+				continue
+			}
+			if err := t.applyECRule(signer, obj, j, payloadParts, nodeLists[j]); err != nil {
+				return fmt.Errorf("apply EC rule #%d (%s): %w", j, ecRules[j], err)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (t *distributedTarget) applyECRule(signer neofscrypto.Signer, obj object.Object, ruleIdx int, payloadParts [][]byte, nodeList []netmap.NodeInfo) error {
+	for partIdx := range payloadParts {
+		// TODO: each part is handled independently, so this worth concurrent execution. https://github.com/nspcc-dev/neofs-node/issues/3504
+		//  Note that distributeTarget.distributeObject is not thread-safe.
+		if err := t.formAndSaveObjectForECPart(signer, obj, ruleIdx, partIdx, payloadParts, nodeList); err != nil {
+			return fmt.Errorf("form and save object for part %d: %w", partIdx, err)
+		}
+	}
+
+	return nil
+}
+
+func (t *distributedTarget) formAndSaveObjectForECPart(signer neofscrypto.Signer, obj object.Object, ruleIdx, partIdx int, payloadParts [][]byte, nodeList []netmap.NodeInfo) error {
+	partObj, err := iec.FormObjectForECPart(signer, obj, payloadParts[partIdx], iec.PartInfo{
+		RuleIndex: ruleIdx,
+		Index:     partIdx,
+	})
+	if err != nil {
+		return fmt.Errorf("form object for part: %w", err)
+	}
+
+	var encObj encodedObject
+	// similar to pkg/services/object/put/distributed.go:95
+	if t.localNodeInContainer {
+		payloadLen := partObj.PayloadSize()
+		if payloadLen > math.MaxInt {
+			return fmt.Errorf("too big payload of physically stored for this server %d > %d", payloadLen, math.MaxInt)
+		}
+
+		hdr := partObj
+		hdr.SetPayload(nil)
+
+		if t.localOnly {
+			encObj, err = encodeObjectWithoutPayload(hdr, int(payloadLen))
+		} else {
+			encObj, err = encodeReplicateRequestWithoutPayload(t.localNodeSigner, hdr, int(payloadLen), t.metainfoConsistencyAttr != "")
+		}
+		if err != nil {
+			return fmt.Errorf("encode object into binary: %w", err)
+		}
+
+		defer putPayload(encObj.b)
+
+		encObj.b = append(encObj.b, partObj.Payload()...)
+	}
+
+	if err := t.saveECPart(partObj, objectcore.ContentMeta{}, encObj, partIdx, len(payloadParts), nodeList); err != nil {
+		return fmt.Errorf("save part object: %w", err)
+	}
+
+	return nil
+}
+
+func (t *distributedTarget) saveECPart(part object.Object, objMeta objectcore.ContentMeta, encObj encodedObject, idx, total int, nodeList []netmap.NodeInfo) error {
+	return t.distributeObject(part, objMeta, encObj, func(obj object.Object, objMeta objectcore.ContentMeta, encObj encodedObject) error {
+		return t.distributeECPart(obj, objMeta, encObj, idx, total, nodeList)
+	})
+}
+
+func (t *distributedTarget) distributeECPart(part object.Object, objMeta objectcore.ContentMeta, enc encodedObject, idx, total int, nodeList []netmap.NodeInfo) error {
+	var firstErr error
+	for {
+		err := t.saveECPartOnNode(part, objMeta, enc, nodeList[idx])
```

Comment:
How about saving on backup nodes (from `nodeList[DATA+PARITY:]` range) if this fails?

Reply:
worthwile addition